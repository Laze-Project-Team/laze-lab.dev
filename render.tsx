import type {
  Queries,
  queries,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { render as defaultRender } from '@testing-library/react';
import type { FC, ReactNode } from 'react';

import { ColorModeProvider } from '@/components/contexts/ColorModeContext';

export type renderType<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container,
> = (
  ui: React.ReactElement,
  options?: RenderOptions<Q, Container, BaseElement>,
) => RenderResult<Q, Container, BaseElement>;

type wrapperProps = {
  children?: ReactNode;
};

type globalWrapperHOCType = (
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>,
) => FC<wrapperProps>;

const globalWrapperHOC: globalWrapperHOCType = (Wrapper) => {
  const globalWrapper: FC<wrapperProps> = ({ children }) => (
    <ColorModeProvider>
      {Wrapper ? (
        <Wrapper>
          <>{children}</>
        </Wrapper>
      ) : (
        children
      )}
    </ColorModeProvider>
  );

  return globalWrapper;
};

export const render: renderType = (ui, options) =>
  defaultRender(ui, {
    ...options,
    wrapper: globalWrapperHOC(options?.wrapper),
  });
