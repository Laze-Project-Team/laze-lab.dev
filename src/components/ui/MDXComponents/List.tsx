import { css } from '@emotion/react';
import type { FC } from 'react';

import { useColorMode } from '@/components/contexts/ColorModeContext';

const listStyle = css`
  padding: 0 0 0 2rem;
  margin: 0.25rem;
  font-size: 0.9rem;
`;

export const Ul: FC<JSX.IntrinsicElements['ul']> = ({ children }) => (
  <ul css={listStyle}>{children}</ul>
);

export const Ol: FC<JSX.IntrinsicElements['ol']> = ({ children }) => (
  <ol css={listStyle}>{children}</ol>
);

export const Li: FC<JSX.IntrinsicElements['li']> = ({ children }) => {
  const { themePattern } = useColorMode();
  const color = themePattern('rgba(0, 0, 0, 0.87)', 'rgb(189, 189, 189)');

  return (
    <li
      css={css`
        padding: 0.25rem 0;
        color: ${color};
      `}
    >
      {children}
    </li>
  );
};
