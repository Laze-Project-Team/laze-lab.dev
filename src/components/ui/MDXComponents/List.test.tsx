import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import { expect, it } from 'vitest';

import { render } from '@/lib/test/render';

import { Li, Ol, Ul } from './List';

const unorderedList = (
  <Ul>
    <Li>item 1</Li>
    <Li>item 2</Li>
    <Li>item 3</Li>
  </Ul>
);

const orderedList = (
  <Ol>
    <Li>item 1</Li>
    <Li>item 2</Li>
    <Li>item 3</Li>
  </Ol>
);

it.each([
  ['ul', unorderedList],
  ['ol', orderedList],
])('should be rendered correctly (%s)', (name, component) => {
  const { container } = render(component);
  expect(container.firstChild).toMatchSnapshot();
});

it.each([
  ['ul', unorderedList],
  ['ol', orderedList],
])('should have "list" role (%s)', (name, component) => {
  render(component);
  expect(screen.getByRole('list')).toBeInTheDocument();
});
