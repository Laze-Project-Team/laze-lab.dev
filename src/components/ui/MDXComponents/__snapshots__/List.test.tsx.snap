// Vitest Snapshot v1

exports[`should be rendered correctly (ol) 1`] = `
<ol
  class="css-ygswl4-listStyle"
>
  <li
    class="css-x78bqw-Li"
  >
    item 1
  </li>
  <li
    class="css-x78bqw-Li"
  >
    item 2
  </li>
  <li
    class="css-x78bqw-Li"
  >
    item 3
  </li>
</ol>
`;

exports[`should be rendered correctly (ul) 1`] = `
<ul
  class="css-ygswl4-listStyle"
>
  <li
    class="css-x78bqw-Li"
  >
    item 1
  </li>
  <li
    class="css-x78bqw-Li"
  >
    item 2
  </li>
  <li
    class="css-x78bqw-Li"
  >
    item 3
  </li>
</ul>
`;
